tests/unittests/CoreGraphics.drawing/CGContextDrawing_ImageMaskTests.cpp

```diff
@@ -0,0 +1,314 @@
+    CGContextRef context = bitmapContext.get();
+
+    drawFunction(context, size, clipType);
+
+    woc::unique_cf<CGImageRef> maskImage{ CGBitmapContextCreateImage(context) };
+
+    if (clipType == ClippingTypeMask) {
+        maskImage.reset(CGImageMaskCreate(CGImageGetWidth(maskImage.get()),
+                                          CGImageGetHeight(maskImage.get()),
+                                          CGImageGetBitsPerComponent(maskImage.get()),
+                                          CGImageGetBitsPerPixel(maskImage.get()),
+                                          CGImageGetBytesPerRow(maskImage.get()),
+                                          CGImageGetDataProvider(maskImage.get()),
+                                          nullptr,
+                                          true));
+    }
+
+    return maskImage.release();
+}
+
+auto __MakeSquareMaskGenerator(CGAffineTransform transform) {
+    // All this work to create a square/diamond with a diamond hole in the middle.
+    return [transform](CGContextRef context, CGSize size, ClippingType clipType) {
+        if (clipType == ClippingTypeMask) {
+            // Since it's a mask, flood it with white.
+            CGContextSetRGBFillColor(context, 1.0, 1.0, 1.0, 1.0);
+            CGContextFillRect(context, { CGPointZero, size });
+        }
+
+        CGContextTranslateCTM(context, size.width / 2.f, size.height / 2.f);
+        CGContextConcatCTM(context, transform);
+        CGContextTranslateCTM(context, -(size.width / 2.f), -(size.height / 2.f));
+
+        CGContextBeginPath(context);
+
+        // Add a set of nested diamonds and clip to them. They should be anchored at 1/4 and 2/3 of the way through the context,
+        // respectively.
+        CGContextAddRect(context, { size.width / 4.f, size.height / 4.f, size.width / 2.f, size.height / 2.f });
+        CGContextAddRect(context, { size.width * .75 / 2.f, size.height * .75 / 2.f, size.width / 4.f, size.height / 4.f });
+        CGContextEOClip(context);
+
+        CGContextSetRGBFillColor(context, 0.5, 0.5, 0.5, clipType == ClippingTypeAlpha ? 0.5 : 1.0);
+        CGContextFillRect(context, { CGPointZero, size });
+
+        // Add an additional full-brightness region (the bottom tip of the diamond.)
+        CGContextSaveGState(context);
+        CGContextAddRect(context, { CGPointZero, size.width / 2.f, size.height / 2.f });
+        CGContextClip(context);
+
+        CGContextSetRGBFillColor(context, 0.0, 0.0, 0.0, 1.0);
+        CGContextFillRect(context, { CGPointZero, size });
+    };
+}
+
+void __GradientMaskGenerator(CGContextRef context, CGSize size, ClippingType clipType) {
+    for (size_t i = 0; i < 256; ++i) {
+        float val = (float)i / 255.f;
+        CGContextSetRGBStrokeColor(context, val, val, val, clipType == ClippingTypeAlpha ? (1.0 - val) : 1.0);
+        CGPoint line[]{
+            { i + .5, 0 }, { i + .5, size.height },
```

Comment:
>5 [](start = 19, length = 1)

nit, .5's should be a constant as they're used so often. This way the .5 can be explained as aligning properly to the pixel boundaries in a single spot.